File: LibraryManagementSystem/code-on-demand/DynamicLoader.py
"""
RESTful Constraint: Code-On-Demand (Optional)

This demonstrates the code-on-demand constraint - the only optional constraint in REST.
The server can extend client functionality by transferring executable code.

Examples:
- JavaScript sent to web browsers
- Applets, plugins, scripts
- Dynamic UI components
- Client-side validation logic

Key Benefits:
- Reduces client complexity
- Server can push updates to client logic
- Improves extensibility
- Reduces need for pre-implementation of features

Note: This is OPTIONAL - many REST APIs don't use it.
"""

from flask import Flask, jsonify, request, render_template_string, send_from_directory
from flask_sqlalchemy import SQLAlchemy
from datetime import datetime
import os

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///library_cod.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)

class Book(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(100), nullable=False)
    author = db.Column(db.String(100), nullable=False)
    isbn = db.Column(db.String(13), unique=True, nullable=False)
    available = db.Column(db.Boolean, default=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'title': self.title,
            'author': self.author,
            'isbn': self.isbn,
            'available': self.available
        }

# ============================================================================
# EXAMPLE 1: Server sends JavaScript for client-side validation
# ============================================================================

@app.route('/api/validation/isbn.js', methods=['GET'])
def get_isbn_validator():
    """
    CODE-ON-DEMAND: Server sends JavaScript validation logic
    Client executes this code for ISBN validation
    Server can update validation rules without client updates
    """
    javascript_code = """
// ISBN Validation Logic - Delivered On-Demand from Server
// Server can update this logic without client changes

function validateISBN(isbn) {
    // Remove hyphens and spaces
    isbn = isbn.replace(/[-\\s]/g, '');
    
    // Check length
    if (isbn.length !== 10 && isbn.length !== 13) {
        return {
            valid: false,
            message: 'ISBN must be 10 or 13 characters'
        };
    }
    
    // Validate ISBN-10
    if (isbn.length === 10) {
        return validateISBN10(isbn);
    }
    
    // Validate ISBN-13
    if (isbn.length === 13) {
        return validateISBN13(isbn);
    }
}

function validateISBN10(isbn) {
    let sum = 0;
    for (let i = 0; i < 9; i++) {
        if (!/\\d/.test(isbn[i])) {
            return { valid: false, message: 'ISBN-10 must contain only digits (except last char can be X)' };
        }
        sum += parseInt(isbn[i]) * (10 - i);
    }
    
    let checkDigit = isbn[9].toUpperCase();
    if (checkDigit === 'X') {
        sum += 10;
    } else if (/\\d/.test(checkDigit)) {
        sum += parseInt(checkDigit);
    } else {
        return { valid: false, message: 'Invalid check digit' };
    }
    
    if (sum % 11 === 0) {
        return { valid: true, message: 'Valid ISBN-10' };
    } else {
        return { valid: false, message: 'Invalid ISBN-10 checksum' };
    }
}

function validateISBN13(isbn) {
    let sum = 0;
    for (let i = 0; i < 13; i++) {
        if (!/\\d/.test(isbn[i])) {
            return { valid: false, message: 'ISBN-13 must contain only digits' };
        }
        let digit = parseInt(isbn[i]);
        sum += (i % 2 === 0) ? digit : digit * 3;
    }
    
    if (sum % 10 === 0) {
        return { valid: true, message: 'Valid ISBN-13' };
    } else {
        return { valid: false, message: 'Invalid ISBN-13 checksum' };
    }
}

// Export for use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { validateISBN };
}

console.log('[CODE-ON-DEMAND] ISBN validation logic loaded from server');
"""
    
    response = app.response_class(
        response=javascript_code,
        mimetype='application/javascript'
    )
    # Allow caching of validation code
    response.headers['Cache-Control'] = 'public, max-age=3600'
    return response


# ============================================================================
# EXAMPLE 2: Server sends JavaScript widget/component
# ============================================================================

@app.route('/api/widgets/book-card.js', methods=['GET'])
def get_book_card_widget():
    """
    CODE-ON-DEMAND: Server sends UI component code
    Client can render books using server-provided logic
    """
    javascript_code = """
// Book Card Widget - Delivered On-Demand
// Server controls presentation logic

class BookCard {
    constructor(book, containerId) {
        this.book = book;
        this.container = document.getElementById(containerId);
        this.render();
    }
    
    render() {
        const card = document.createElement('div');
        card.className = 'book-card';
        card.style.cssText = `
            border: 1px solid #ddd;
            border-radius: 8px;
            padding: 16px;
            margin: 10px;
            background: white;
            box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        `;
        
        card.innerHTML = `
            <h3 style="margin: 0 0 10px 0; color: #333;">${this.book.title}</h3>
            <p style="margin: 5px 0; color: #666;">
                <strong>Author:</strong> ${this.book.author}
            </p>
            <p style="margin: 5px 0; color: #666;">
                <strong>ISBN:</strong> ${this.book.isbn}
            </p>
            <p style="margin: 5px 0;">
                <span class="status" style="
                    padding: 4px 8px;
                    border-radius: 4px;
                    font-size: 12px;
                    font-weight: bold;
                    background: ${this.book.available ? '#4CAF50' : '#f44336'};
                    color: white;
                ">
                    ${this.book.available ? 'Available' : 'Borrowed'}
                </span>
            </p>
            <button onclick="alert('Book ID: ${this.book.id}')" style="
                margin-top: 10px;
                padding: 8px 16px;
                background: #2196F3;
                color: white;
                border: none;
                border-radius: 4px;
                cursor: pointer;
            ">
                View Details
            </button>
        `;
        
        this.container.appendChild(card);
    }
}

// Function to render all books
function renderBooks(books, containerId) {
    const container = document.getElementById(containerId);
    container.innerHTML = '';
    books.forEach(book => new BookCard(book, containerId));
}

console.log('[CODE-ON-DEMAND] Book card widget loaded from server');
"""
    
    response = app.response_class(
        response=javascript_code,
        mimetype='application/javascript'
    )
    response.headers['Cache-Control'] = 'public, max-age=3600'
    return response


# ============================================================================
# EXAMPLE 3: Interactive HTML client that loads code on-demand
# ============================================================================

@app.route('/', methods=['GET'])
def index():
    """
    HTML client that demonstrates code-on-demand
    Dynamically loads JavaScript from server
    """
    html = """
<!DOCTYPE html>
<html>
<head>
    <title>Code-On-Demand Demo - Library</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            max-width: 1200px;
            margin: 0 auto;
            padding: 20px;
            background: #f5f5f5;
        }
        .header {
            background: #2196F3;
            color: white;
            padding: 20px;
            border-radius: 8px;
            margin-bottom: 20px;
        }
        .section {
            background: white;
            padding: 20px;
            border-radius: 8px;
            margin-bottom: 20px;
            box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        }
        input, button {
            padding: 10px;
            margin: 5px;
            border: 1px solid #ddd;
            border-radius: 4px;
        }
        button {
            background: #4CAF50;
            color: white;
            border: none;
            cursor: pointer;
        }
        button:hover {
            background: #45a049;
        }
        .validation-result {
            margin-top: 10px;
            padding: 10px;
            border-radius: 4px;
        }
        .valid {
            background: #dff0d8;
            color: #3c763d;
        }
        .invalid {
            background: #f2dede;
            color: #a94442;
        }
        .info {
            background: #d9edf7;
            padding: 10px;
            border-radius: 4px;
            margin: 10px 0;
        }
    </style>
</head>
<body>
    <div class="header">
        <h1>📚 Code-On-Demand Demonstration</h1>
        <p>This page dynamically loads code from the server</p>
    </div>

    <div class="section">
        <h2>Example 1: Dynamic Validation Logic</h2>
        <div class="info">
            <strong>Code-On-Demand:</strong> ISBN validation logic is loaded from server.
            The server can update validation rules without changing the client.
        </div>
        <input type="text" id="isbnInput" placeholder="Enter ISBN (e.g., 0132350882)" />
        <button onclick="validateISBN()">Validate ISBN</button>
        <div id="validationResult"></div>
    </div>

    <div class="section">
        <h2>Example 2: Dynamic UI Components</h2>
        <div class="info">
            <strong>Code-On-Demand:</strong> Book display widget is loaded from server.
            The server controls how books are presented.
        </div>
        <button onclick="loadBooks()">Load Books</button>
        <div id="booksContainer"></div>
    </div>

    <div class="section">
        <h2>Code Loading Status</h2>
        <div id="codeStatus"></div>
    </div>

    <script>
        // Step 1: Load ISBN validation code from server (Code-On-Demand)
        console.log('[CLIENT] Loading ISBN validation code from server...');
        const validationScript = document.createElement('script');
        validationScript.src = '/api/validation/isbn.js';
        validationScript.onload = () => {
            document.getElementById('codeStatus').innerHTML += 
                '<p style="color: green;">✓ ISBN validation code loaded from server</p>';
        };
        document.head.appendChild(validationScript);

        // Step 2: Load book card widget from server (Code-On-Demand)
        console.log('[CLIENT] Loading book widget code from server...');
        const widgetScript = document.createElement('script');
        widgetScript.src = '/api/widgets/book-card.js';
        widgetScript.onload = () => {
            document.getElementById('codeStatus').innerHTML += 
                '<p style="color: green;">✓ Book widget code loaded from server</p>';
        };
        document.head.appendChild(widgetScript);

        // Client-side function to validate ISBN using server-provided code
        function validateISBN() {
            const isbn = document.getElementById('isbnInput').value;
            const resultDiv = document.getElementById('validationResult');
            
            // Use validation function loaded from server
            const result = window.validateISBN(isbn);
            
            resultDiv.className = 'validation-result ' + (result.valid ? 'valid' : 'invalid');
            resultDiv.innerHTML = `
                <strong>${result.valid ? '✓' : '✗'} ${result.message}</strong>
                <p>Validation performed using server-provided logic</p>
            `;
        }

        // Client-side function to load and display books
        async function loadBooks() {
            const response = await fetch('/api/books');
            const data = await response.json();
            
            // Use widget loaded from server to render books
            window.renderBooks(data.data, 'booksContainer');
        }

        // Display info about code-on-demand
        document.getElementById('codeStatus').innerHTML += 
            '<p><strong>Note:</strong> This page loaded executable code from the server:</p>' +
            '<ul>' +
            '<li>/api/validation/isbn.js - Validation logic</li>' +
            '<li>/api/widgets/book-card.js - UI component</li>' +
            '</ul>' +
            '<p>The server can update this code without changing the client!</p>';
    </script>
</body>
</html>
    """
    return render_template_string(html)


# ============================================================================
# Standard API endpoints
# ============================================================================

@app.route('/api/books', methods=['GET'])
def get_books():
    """Get all books"""
    books = Book.query.all()
    return jsonify({
        'success': True,
        'data': [book.to_dict() for book in books]
    })


@app.route('/api/books', methods=['POST'])
def create_book():
    """Create a new book"""
    data = request.get_json()
    
    book = Book(
        title=data['title'],
        author=data['author'],
        isbn=data['isbn']
    )
    
    db.session.add(book)
    db.session.commit()
    
    return jsonify({
        'success': True,
        'data': book.to_dict(),
        'message': 'Book created'
    }), 201


if __name__ == '__main__':
    with app.app_context():
        db.create_all()
        
        # Add sample data
        if Book.query.count() == 0:
            samples = [
                Book(title="Designing Data-Intensive Applications", author="Martin Kleppmann", isbn="1449373321"),
                Book(title="Site Reliability Engineering", author="Google", isbn="149192912X"),
                Book(title="The Phoenix Project", author="Gene Kim", isbn="0988262592"),
            ]
            db.session.add_all(samples)
            db.session.commit()
            print("✓ Sample books added")
    
    print("="*70)
    print("CODE-ON-DEMAND DEMONSTRATION")
    print("="*70)
    print("Server running on http://127.0.0.1:5006")
    print("\nCode-On-Demand Features:")
    print("  • Server sends JavaScript validation logic")
    print("  • Server sends UI component code")
    print("  • Client executes server-provided code")
    print("  • Server can update logic without client changes")
    print("\nThis is the ONLY OPTIONAL REST constraint!")
    print("\nOpen browser: http://127.0.0.1:5006")
    print("="*70)
    
    app.run(debug=True, port=5006)

File: LibraryManagementSystem/code-on-demand/DynamicLoader.js
/**
 * CODE-ON-DEMAND CONSTRAINT (Optional)
 * 
 * This is the only OPTIONAL constraint in REST architecture.
 * Servers can temporarily extend client functionality by transferring executable code.
 * 
 * Benefits:
 * - Improved client extensibility
 * - Reduced complexity by moving features to server-side
 * - Easy deployment of features (no client updates needed)
 * 
 * Trade-offs:
 * - Reduces visibility (code executed may not be inspectable)
 * - Security concerns (executing untrusted code)
 * 
 * Common examples:
 * - JavaScript sent from server to browser
 * - Applets, plugins
 * - Server-side rendered components
 */

// ==================== SERVER-SIDE: Code Provider ====================

class CodeOnDemandServer {
    /**
     * Server provides validation logic to client on-demand
     * Client can request and execute this code dynamically
     */
    static getValidationCode() {
        // Server-side validation rules that can be sent to client
        return `
            // ISBN Validation (dynamically loaded from server)
            function validateISBN(isbn) {
                // Remove hyphens and spaces
                const cleaned = isbn.replace(/[-\\s]/g, '');
                
                // Check if it's 13 digits
                if (!/^\\d{13}$/.test(cleaned)) {
                    return {
                        valid: false,
                        message: 'ISBN must be 13 digits'
                    };
                }
                
                // Calculate check digit
                let sum = 0;
                for (let i = 0; i < 12; i++) {
                    const digit = parseInt(cleaned[i]);
                    sum += (i % 2 === 0) ? digit : digit * 3;
                }
                const checkDigit = (10 - (sum % 10)) % 10;
                
                if (checkDigit !== parseInt(cleaned[12])) {
                    return {
                        valid: false,
                        message: 'Invalid ISBN check digit'
                    };
                }
                
                return {
                    valid: true,
                    message: 'Valid ISBN'
                };
            }
            
            // Email Validation (dynamically loaded)
            function validateEmail(email) {
                const pattern = /^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$/;
                return {
                    valid: pattern.test(email),
                    message: pattern.test(email) ? 'Valid email' : 'Invalid email format'
                };
            }
            
            // Borrowing period validation
            function validateBorrowDays(days) {
                if (days < 1 || days > 30) {
                    return {
                        valid: false,
                        message: 'Borrow period must be between 1 and 30 days'
                    };
                }
                return {
                    valid: true,
                    message: 'Valid borrow period'
                };
            }
        `;
    }

    /**
     * Server provides UI enhancement code on-demand
     */
    static getUIEnhancementCode() {
        return `
            // Dynamic book availability indicator
            function enhanceBookDisplay(bookElement, bookData) {
                const indicator = document.createElement('div');
                indicator.className = 'availability-indicator';
                
                if (bookData.available) {
                    indicator.innerHTML = '✓ Available';
                    indicator.style.color = 'green';
                    
                    // Add borrow button
                    const borrowBtn = document.createElement('button');
                    borrowBtn.textContent = 'Borrow Now';
                    borrowBtn.onclick = () => initiateBookBorrow(bookData.id);
                    bookElement.appendChild(borrowBtn);
                } else {
                    indicator.innerHTML = '✗ Currently Borrowed';
                    indicator.style.color = 'red';
                    
                    // Add notification button
                    const notifyBtn = document.createElement('button');
                    notifyBtn.textContent = 'Notify When Available';
                    notifyBtn.onclick = () => subscribeToAvailability(bookData.id);
                    bookElement.appendChild(notifyBtn);
                }
                
                bookElement.appendChild(indicator);
            }
            
            // Auto-complete for book search
            function enableSearchAutocomplete(inputElement) {
                let timeout;
                inputElement.addEventListener('input', function() {
                    clearTimeout(timeout);
                    timeout = setTimeout(() => {
                        const query = this.value;
                        if (query.length >= 3) {
                            fetch('/api/books?search=' + query)
                                .then(res => res.json())
                                .then(data => showSuggestions(data.books));
                        }
                    }, 300);
                });
            }
        `;
    }

    /**
     * Server provides analytics code on-demand
     */
    static getAnalyticsCode() {
        return `
            // Track user interactions
            function trackBookView(bookId) {
                fetch('/api/analytics/book-view', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ book_id: bookId, timestamp: Date.now() })
                });
            }
            
            function trackSearch(query, resultCount) {
                fetch('/api/analytics/search', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ 
                        query: query, 
                        results: resultCount,
                        timestamp: Date.now() 
                    })
                });
            }
        `;
    }

    /**
     * API endpoint to serve code on-demand
     */
    static handleCodeRequest(req, res) {
        const codeType = req.params.type;
        
        let code;
        switch (codeType) {
            case 'validation':
                code = this.getValidationCode();
                break;
            case 'ui-enhancement':
                code = this.getUIEnhancementCode();
                break;
            case 'analytics':
                code = this.getAnalyticsCode();
                break;
            default:
                res.status(404).send({ error: 'Code type not found' });
                return;
        }
        
        res.set('Content-Type', 'application/javascript');
        res.send(code);
    }
}

// ==================== CLIENT-SIDE: Code Consumer ====================

class DynamicCodeLoader {
    constructor() {
        this.loadedModules = new Map();
    }

    /**
     * Load and execute code from server dynamically
     */
    async loadCode(codeType) {
        console.log(`[CLIENT] Requesting ${codeType} code from server...`);
        
        // Check if already loaded
        if (this.loadedModules.has(codeType)) {
            console.log(`[CLIENT] Code already loaded: ${codeType}`);
            return this.loadedModules.get(codeType);
        }

        try {
            // Request code from server
            const response = await fetch(`/api/code/${codeType}`);
            const code = await response.text();
            
            console.log(`[CLIENT] Received code from server (${code.length} bytes)`);
            
            // Execute the code
            // In browser: creates functions in global scope
            // In Node.js: eval in local scope
            eval(code);
            
            console.log(`[CLIENT] Code executed successfully`);
            
            this.loadedModules.set(codeType, true);
            return true;
        } catch (error) {
            console.error(`[CLIENT] Error loading code: ${error.message}`);
            return false;
        }
    }

    /**
     * Example: Load validation code and use it
     */
    async enhanceFormValidation(formElement) {
        // Load validation code from server
        await this.loadCode('validation');
        
        // Now we can use the dynamically loaded validation functions
        const isbnInput = formElement.querySelector('#isbn');
        const emailInput = formElement.querySelector('#email');
        const daysInput = formElement.querySelector('#days');
        
        // Add validators (functions loaded from server)
        isbnInput.addEventListener('blur', function() {
            const result = validateISBN(this.value); // Function loaded from server!
            showValidationResult(this, result);
        });
        
        emailInput.addEventListener('blur', function() {
            const result = validateEmail(this.value);
            showValidationResult(this, result);
        });
        
        daysInput.addEventListener('blur', function() {
            const result = validateBorrowDays(parseInt(this.value));
            showValidationResult(this, result);
        });
        
        console.log('[CLIENT] Form validation enhanced with server-provided code');
    }

    /**
     * Example: Load UI enhancement code
     */
    async enhanceBookCatalog(catalogElement) {
        await this.loadCode('ui-enhancement');
        
        // Use server-provided UI enhancement
        const books = catalogElement.querySelectorAll('.book-item');
        books.forEach(bookEl => {
            const bookData = JSON.parse(bookEl.dataset.book);
            enhanceBookDisplay(bookEl, bookData); // Function loaded from server!
        });
        
        // Enable search autocomplete
        const searchInput = document.querySelector('#book-search');
        if (searchInput) {
            enableSearchAutocomplete(searchInput); // Function loaded from server!
        }
        
        console.log('[CLIENT] Book catalog enhanced with server-provided code');
    }

    /**
     * Example: Load analytics code
     */
    async enableAnalytics() {
        await this.loadCode('analytics');
        
        // Use server-provided analytics
        document.querySelectorAll('.book-item').forEach(bookEl => {
            bookEl.addEventListener('click', function() {
                const bookId = this.dataset.bookId;
                trackBookView(bookId); // Function loaded from server!
            });
        });
        
        const searchForm = document.querySelector('#search-form');
        if (searchForm) {
            searchForm.addEventListener('submit', function(e) {
                const query = this.querySelector('input').value;
                const results = document.querySelectorAll('.search-result').length;
                trackSearch(query, results); // Function loaded from server!
            });
        }
        
        console.log('[CLIENT] Analytics enabled with server-provided code');
    }
}

// Helper function for validation display
function showValidationResult(inputElement, result) {
    const feedback = inputElement.nextElementSibling || document.createElement('div');
    feedback.className = 'validation-feedback';
    feedback.textContent = result.message;
    feedback.style.color = result.valid ? 'green' : 'red';
    
    if (!inputElement.nextElementSibling) {
        inputElement.parentNode.insertBefore(feedback, inputElement.nextSibling);
    }
}

// ==================== DEMONSTRATION ====================

async function demonstrateCodeOnDemand() {
    console.log('=== Code-On-Demand Constraint Demo ===\n');
    
    const loader = new DynamicCodeLoader();
    
    // Scenario 1: Load validation code on-demand
    console.log('\n--- Scenario 1: Dynamic Form Validation ---');
    await loader.loadCode('validation');
    
    // Test the dynamically loaded validation
    console.log('\nTesting ISBN validation (code from server):');
    // Note: validateISBN function is now available!
    console.log('  9780132350884:', validateISBN('9780132350884'));
    console.log('  1234567890123:', validateISBN('1234567890123'));
    
    // Scenario 2: UI Enhancement code
    console.log('\n--- Scenario 2: Dynamic UI Enhancement ---');
    console.log('Loading UI enhancement code from server...');
    await loader.loadCode('ui-enhancement');
    console.log('UI enhancement code loaded and ready to use');
    
    // Scenario 3: Analytics code
    console.log('\n--- Scenario 3: Dynamic Analytics ---');
    await loader.loadCode('analytics');
    console.log('Analytics code loaded from server');
    
    console.log('\n=== Benefits of Code-On-Demand ===');
    console.log('✓ Client receives up-to-date validation rules from server');
    console.log('✓ New features deployed without client updates');
    console.log('✓ Reduced initial client complexity');
    console.log('✓ Server controls business logic centrally');
    
    console.log('\n=== Trade-offs ===');
    console.log('⚠ Reduced visibility (executed code may not be inspectable)');
    console.log('⚠ Security concerns (must trust server)');
    console.log('⚠ Network dependency (need code before execution)');
}

// Run demonstration in browser or Node.js
if (typeof window === 'undefined') {
    // Node.js environment
    console.log('Code-On-Demand demonstration (simulated)');
    console.log('In a browser, this would dynamically load and execute server code\n');
    
    // Simulate server code
    eval(CodeOnDemandServer.getValidationCode());
    
    console.log('Example: Validating ISBN with server-provided code');
    console.log(validateISBN('9780132350884'));
} else {
    // Browser environment
    demonstrateCodeOnDemand();
}

// ==================== EXPORT ====================

if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        CodeOnDemandServer,
        DynamicCodeLoader
    };
}
